Integrate fault localizer into goto checker [blocks: 4215]

peterschrammel opened the pull request:
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Use goto verifier for fault localization in CBMC

Available by using multi-path symex checker

peterschrammel committed:
Add freeze_guards to bmc_util

Utility function to freeze all guards, as e.g.
needed for fault localization.

peterschrammel committed:
Make multi-path symex checker a fault localization provider

Uses goto_symex_fault_localizert to provide the functionality.

peterschrammel committed:
Add const version for returning last trace step

peterschrammel committed:
Add fault localization output functions

Print information about the likely fault location.

peterschrammel committed:
Add fault localization provider interface

Can be implemented by an incremental goto checker
to provide information about the likely fault location.

peterschrammel committed:
Fault localizer for goto-symex

Requires prop_convt, symex_target_equationt and
frozen guards.

peterschrammel committed:
All properties verifier with fault localization

Variant that reports fault localization information.

peterschrammel committed:
Stop-on-fail verifier with fault localization

Variant that reports fault localization information.

peterschrammel committed:
Use goto verifier for fault localization in JBMC

Available by using multi-path symex checker

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eca95300d4183112f21902ddb4b5b1beacdc67c0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101371546

tautschnig commented on src/goto-checker/report_util.cpp:
```diff
@@ -362,6 +367,256 @@ void output_properties_with_traces(
+  {
+    log.debug() << score_pair.first->source_location
+                << "\n  score: " << score_pair.second << messaget::eom;
+  }
```
Use `messaget::conditional_output` to avoid doing a lot of work when it most likely won't be of any value.

tautschnig commented on src/goto-checker/goto_symex_fault_localizer.cpp:
```diff
@@ -0,0 +1,146 @@
+    log.status() << "Localizing fault" << messaget::eom;
+
+    // pick localization method
+    //  if(options.get_option("localize-faults-method")=="TBD")
```
I suppose this is copy&paste?

peterschrammel replied:
Yes, I am inclined to remove this option altogether as nobody has ever implemented any alternative method in 3 years.

tautschnig commented on src/goto-checker/goto_symex_fault_localizer.cpp:
```diff
@@ -0,0 +1,146 @@
+  if(solver() == decision_proceduret::resultt::D_SATISFIABLE)
+    return false;
+
+  return true;
```
`return solver() != decision_proceduret::resultt::D_SATISFIABLE;`

tautschnig commented on src/goto-checker/goto_symex_fault_localizer.cpp:
```diff
@@ -0,0 +1,146 @@
+  v.resize(localization_points.size());
+
+  for(std::size_t i = 0; i < localization_points.size(); ++i)
+    v[i] = tvt(tvt::tv_enumt::TV_UNKNOWN);
```
`localization_points_valuet v(localization_points.size(), tvt::unknown());`

smowton commented on src/goto-checker/report_util.cpp:
```diff
@@ -362,6 +367,256 @@ void output_properties_with_traces(
+{
+  PRECONDITION(!fault_location.scores.empty());
+
+  auto max_it = fault_location.scores.begin();
```
Use `std::max` and a comparator?

smowton commented on src/goto-checker/goto_symex_fault_localizer.cpp:
```diff
@@ -0,0 +1,146 @@
+    {
+      l.second->second++;
+    }
+    else if(solver.l_get(l.first).is_false() && l.second->second > 0)
```
`l.second->second` is a bit opaque -- comment on what that is?

smowton commented on src/goto-checker/all_properties_verifier_with_fault_localization.h:
```diff
@@ -0,0 +1,109 @@
+        if(properties.at(property_id).status == property_statust::FAIL)
+        {
+          // get correctly truncated error trace for property and store it
+          (void)traces.insert(
```
Are these both expected to be insert-new-key? If so suggest cross-checking with a CHECK_RETURN that they are indeed new.

smowton replied:
(If not, suggest inserting a dummy value and only calling the expensive build / localize functions if they really are new)

peterschrammel replied:
I've added an invariant in the trace storage.

peterschrammel committed:
Remove unused localize-faults-method option

peterschrammel committed:
Add invariant for avoiding duplicated error traces

Make sure a failed property only occurs once in updated_properties
so that storing its trace is only attempted once.
